File: src/lib/server/stage-timing-context.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { randomUUID } from "node:crypto";

/** Telemetry identity only; never an edit-approval or mutation authority. */
export interface StageTimingContext {
  runId: string;
  attemptId: string;
  attemptNo: number;
  parentSpanId?: string;
}

export interface StageTimingMetadata {
  provider?: string;
  model?: string;
  effort?: string;
  phase?: string;
  cache?: "hit" | "miss" | "unproved";
  round?: number;
  packetBytes?: number;
  promptBytes?: number;
  deadlineMs?: number;
  lens?: string;
  evidenceImages?: number;
  exitCode?: number;
}

const context = new AsyncLocalStorage<StageTimingContext>();
const writerId = randomUUID();

/** Scope identity to this async call tree, isolating concurrent projects. */
export function withStageTimingContext<T>(value: StageTimingContext, run: () => T): T {
  return context.run({ ...value }, run);
}

/** Standalone helpers may supply a job identity, but must not overwrite an
 * enclosing execution UUID or sever its parent span when invoked by a stage. */
export function withStageTimingFallback<T>(value: StageTimingContext, run: () => T): T {
  return context.getStore() ? run() : withStageTimingContext(value, run);
}

/** Read stable attempt lineage or explicitly identify standalone execution. */
export function stageTimingContext(): StageTimingContext {
  return context.getStore() ?? {
    runId: `standalone:${writerId}`,
    attemptId: writerId,
    attemptNo: 1,
  };
}

/** A process incarnation is the only domain where monotonic subtraction is valid. */
export function stageTimingWriter() {
  return { writerId, writerPid: process.pid, clock: "process-monotonic" } as const;
}

/** Pass explicit parent context to a child, without mutating global process.env. */
export function stageTimingEnv(): NodeJS.ProcessEnv {
  const value = stageTimingContext();
  return {
    ...process.env,
    SNIPER_TIMING_RUN_ID: value.runId,
    SNIPER_TIMING_ATTEMPT_ID: value.attemptId,
    SNIPER_TIMING_ATTEMPT_NO: String(value.attemptNo),
    SNIPER_TIMING_PARENT_SPAN_ID: value.parentSpanId ?? "",
  };
}

/** Only bounded diagnostic scalars; never prompts, transcripts or error payloads. */
export function timingMetadata(value: StageTimingMetadata = {}): StageTimingMetadata {
  const result: Record<string, string | number> = {};
  for (const key of ["provider", "model", "effort", "phase", "cache", "lens"] as const) {
    const item = value[key];
    if (typeof item === "string" && item.length <= 128) result[key] = item;
  }
  for (const key of ["round", "packetBytes", "promptBytes", "deadlineMs", "evidenceImages", "exitCode"] as const) {
    const item = value[key];
    if (typeof item === "number" && Number.isFinite(item)) result[key] = item;
  }
  return result;
}

File: scripts/producer/stage_timing_context.py
"""Process-local timing lineage shared with the TypeScript controller."""
from __future__ import annotations

import contextvars
import os
import uuid

PARENT_SPAN: contextvars.ContextVar[str | None] = contextvars.ContextVar(
    "sniper_timing_parent_span", default=None)
_WRITER_ID = str(uuid.uuid4())
_TEXT_KEYS = ("provider", "model", "effort", "phase", "cache")
_NUMBER_KEYS = ("round", "packetBytes", "evidenceImages", "exitCode")


def timing_context() -> dict:
    """Resolve explicit child lineage; label standalone execution honestly."""
    attempt = os.environ.get("SNIPER_TIMING_ATTEMPT_NO", "1")
    number = int(attempt) if attempt.isdecimal() and len(attempt) <= 8 else 1
    fields = {
        "schemaVersion": 2,
        "runId": os.environ.get("SNIPER_TIMING_RUN_ID") or f"standalone:{_WRITER_ID}",
        "attemptId": os.environ.get("SNIPER_TIMING_ATTEMPT_ID") or _WRITER_ID,
        "attemptNo": max(1, number),
        "writerId": f"{_WRITER_ID}:{os.getpid()}",
        "writerPid": os.getpid(),
        "clock": "process-monotonic",
    }
    parent = PARENT_SPAN.get() or os.environ.get("SNIPER_TIMING_PARENT_SPAN_ID")
    if parent:
        fields["parentSpanId"] = parent
    return fields


def timing_metadata(value: dict | None) -> dict:
    """Whitelist bounded non-content fields; reject arbitrary diagnostic payloads."""
    source = value or {}
    result = {key: source[key] for key in _TEXT_KEYS
              if isinstance(source.get(key), str) and len(source[key]) <= 128}
    for key in _NUMBER_KEYS:
        item = source.get(key)
        if (isinstance(item, (int, float)) and not isinstance(item, bool)
                and abs(item) <= 2**53 - 1):
            result[key] = item
    return result


def timing_environment() -> dict[str, str]:
    """Forward the current parent span to nested Python subprocesses explicitly."""
    context = timing_context()
    return {**os.environ,
            "SNIPER_TIMING_RUN_ID": context["runId"],
            "SNIPER_TIMING_ATTEMPT_ID": context["attemptId"],
            "SNIPER_TIMING_ATTEMPT_NO": str(context["attemptNo"]),
            "SNIPER_TIMING_PARENT_SPAN_ID": context.get("parentSpanId", "")}
